### DISCUSSION TITLE
Possible error in the default bimodal preditor

### DISCUSSION THREAD

**Discussion by Mingcong-Cao** [ChampSim/ChampSim#452] [General]
The branch prediction condition in branch/bimodal/bimodal.cc is:
```
uint8_t O3_CPU::predict_branch(uint64_t ip)
{
  auto hash = ip % ::BIMODAL_PRIME;
  auto value = ::bimodal_table[this][hash];
  return value.value() >= (value.maximum / 2);
}
```

where value.maximum is 2^n - 1 as far as I understand (n is the COUNTER_BITS). Thus, in order to predict not taken for value < 2^(n-1) and taken for value >= 2^(n-1), shouldn't we use > instead of >= in this condition.

**Comment by ngober** [comment]
I think you're correct. This looks like an integer division bug. Fixed: https://github.com/ChampSim/ChampSim/commit/da69133e86387e546d451e9d7699d9f9f80d30ab
